File: app/src/main/java/com/blanke/mdwechat/ui/BaseHookUi.java
package com.blanke.mdwechat.ui;

import android.app.Activity;
import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.blanke.mdwechat.WeChatHelper;
import com.blanke.mdwechat.config.HookConfig;

import de.robv.android.xposed.callbacks.XC_LoadPackage;

/**
 * Created by blanke on 2017/8/1.
 */

public abstract class BaseHookUi {
    private final String TAG = getClass().getSimpleName();
    private Drawable rippleDrawable;

    public abstract void hook(XC_LoadPackage.LoadPackageParam lpparam);

    //刷新设置
    protected void refreshPrefs() {
        WeChatHelper.XMOD_PREFS.reload();
    }

    protected View findViewByIdName(Activity activity, String idName) {
        return activity.findViewById(getViewId(activity, idName));
    }

    protected View findViewByIdName(View view, String idName) {
        return view.findViewById(getViewId(view.getContext(), idName));
    }

    @Deprecated
    protected int getId(Context context, String idName) {
        return context.getResources().getIdentifier(context.getPackageName() + ":id/" + idName, null, null);
    }

    protected int getViewId(Context context, String name) {
        return getResourceIdByName(context, "id", name);
    }

    protected int getColorId(Context context, String idName) {
        return getResourceIdByName(context, "color", idName);
    }

    protected int getDrawableIdByName(Context context, String name) {
        return getResourceIdByName(context, "drawable", name);
    }

    protected int getResourceIdByName(Context context, String resourceName, String name) {
        return context.getResources().getIdentifier(
                context.getPackageName() + ":" + resourceName + "/" + name,
                null, null);
    }

    protected ColorDrawable getColorPrimaryDrawable() {
        ColorDrawable colorDrawable = new ColorDrawable();
        colorDrawable.setColor(getColorPrimary());
        return colorDrawable;
    }

    protected int getColorPrimary() {
        return HookConfig.getColorPrimary();
    }

    protected Drawable getRippleDrawable(Context context) {
        if (rippleDrawable == null) {
            int[] attrs = new int[]{android.R.attr.selectableItemBackground};
            TypedArray ta = context.obtainStyledAttributes(attrs);
            rippleDrawable = ta.getDrawable(0);
            ta.recycle();
        } else {
            return rippleDrawable.getConstantState().newDrawable().mutate();
        }
        return rippleDrawable;
    }

//    protected Drawable getGreyRippleDrawable(Context context) {
//        return DrawableUtils.getPressedColorRippleDrawable(ContextCompat.getColor(context, R.color.colorBlueGrey)
//                , ContextCompat.getColor(context, R.color.colorGrey));
//    }

    /**
     * 打印调试 view tree
     *
     * @param rootView
     * @param level
     */
    protected void printViewTree(View rootView, int level) {
        logSpace(level, getViewMsg(rootView));
        if (rootView instanceof ViewGroup) {
            ViewGroup vg = (ViewGroup) rootView;
            for (int i = 0; i < vg.getChildCount(); i++) {
                View v = vg.getChildAt(i);
                printViewTree(v, level + 1);
            }
        }
    }

    protected void logViewTree(View view) {
        printViewTree(view, 0);
    }

    protected void printActivityViewTree(Activity activity) {
        View contentView = activity.findViewById(android.R.id.content);
        printViewTree(contentView, 0);
    }

    protected void printActivityWindowViewTree(Activity activity) {
        log("-------start----------");
        printViewTree(activity.getWindow().getDecorView(), 0);
        log("-------end----------");
    }

    protected Drawable getTransparentDrawable() {
        return new ColorDrawable(Color.TRANSPARENT);
    }

    protected Drawable getWhiteDrawable() {
        return new ColorDrawable(Color.WHITE);
    }

    protected String getViewMsg(View view) {
        String className = view.getClass().getName();
        int id = view.getId();
        String text = "";
        if (view instanceof TextView) {
            TextView textView = (TextView) view;
            text = textView.getText().toString();
            text += "(" + textView.getHint() + ")";
        }
        return className + "," + id + "," + text + "," + view;
    }

    private void logSpace(int count, String msg) {
        StringBuilder sb = new StringBuilder();
        for (int j = 0; j < count; j++) {
            sb.append("----");
        }
        sb.append(msg);
        log(sb.toString());
    }

    protected void log(View view) {
        log(getViewMsg(view));
    }

    protected void log(String msg) {
        LogUtil.log(TAG + ":" + msg);
    }

    protected void log(Throwable e) {
        LogUtil.log(e);
    }

    protected void logSuperClass(Class clazz) {
        if (clazz == Object.class) {
            return;
        }
        log(clazz.getName());
        logSuperClass(clazz.getSuperclass());
    }

    protected void logStackTraces() {
        logStackTraces(15);
    }

    protected void logStackTraces(int methodCount) {
        logStackTraces(methodCount, 3);
    }

    protected void logStackTraces(int methodCount, int methodOffset) {
        StackTraceElement[] trace = Thread.currentThread().getStackTrace();
        String level = "";
        log("---------logStackTraces start----------");
        for (int i = methodCount; i > 0; i--) {
            int stackIndex = i + methodOffset;
            if (stackIndex >= trace.length) {
                continue;
            }
            StringBuilder builder = new StringBuilder();
            builder.append("|")
                    .append(' ')
                    .append(level)
                    .append(trace[stackIndex].getClassName())
                    .append(".")
                    .append(trace[stackIndex].getMethodName())
                    .append(" ")
                    .append(" (")
                    .append(trace[stackIndex].getFileName())
                    .append(":")
                    .append(trace[stackIndex].getLineNumber())
                    .append(")");
            level += "   ";
            log(builder.toString());
        }
        log("---------logStackTraces end----------");
    }
}

File: app/src/main/java/com/blanke/mdwechat/ui/BaseHookUi.kt
package com.blanke.mdwechat.ui

import android.app.Activity
import android.content.Context
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import android.support.v4.content.ContextCompat
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.blanke.mdwechat.R
import com.blanke.mdwechat.WeChatHelper
import com.blanke.mdwechat.config.HookConfig
import com.blanke.mdwechat.util.DrawableUtils
import com.blanke.mdwechat.util.LogUtil
import de.robv.android.xposed.callbacks.XC_LoadPackage

/**
 * Created by blanke on 2017/8/1.
 */

abstract class BaseHookUi {
    private val TAG = javaClass.simpleName
    private var rippleDrawable: Drawable? = null

    abstract fun hook(lpparam: XC_LoadPackage.LoadPackageParam)

    //刷新设置
    protected fun refreshPrefs() {
        WeChatHelper.XMOD_PREFS!!.reload()
    }

    protected fun findViewByIdName(activity: Activity, idName: String): View {
        return activity.findViewById(getViewId(activity, idName))
    }

    protected fun findViewByIdName(view: View, idName: String): View {
        return view.findViewById(getViewId(view.context, idName))
    }

    @Deprecated("")
    protected fun getId(context: Context, idName: String): Int {
        return context.resources.getIdentifier(context.packageName + ":id/" + idName, null, null)
    }

    protected fun getViewId(context: Context, name: String): Int {
        return getResourceIdByName(context, "id", name)
    }

    protected fun getColorId(context: Context, idName: String): Int {
        return getResourceIdByName(context, "color", idName)
    }

    protected fun getDrawableIdByName(context: Context, name: String): Int {
        return getResourceIdByName(context, "drawable", name)
    }

    protected fun getResourceIdByName(context: Context, resourceName: String, name: String): Int {
        return context.resources.getIdentifier(
                context.packageName + ":" + resourceName + "/" + name, null, null)
    }

    protected val colorPrimaryDrawable: ColorDrawable
        get() {
            val colorDrawable = ColorDrawable()
            colorDrawable.color = colorPrimary
            return colorDrawable
        }

    protected val colorPrimary: Int
        get() = HookConfig.colorPrimary

    protected fun getDefaultRippleDrawable(context: Context): Drawable {
        if (rippleDrawable == null) {
            val attrs = intArrayOf(android.R.attr.selectableItemBackground)
            val ta = context.obtainStyledAttributes(attrs)
            rippleDrawable = ta.getDrawable(0)
            ta.recycle()
        } else {
            return rippleDrawable!!.constantState!!.newDrawable().mutate()
        }
        return rippleDrawable!!
    }

    protected fun getGreyRippleDrawable(context: Context): Drawable {
        return DrawableUtils.getPressedColorRippleDrawable(Color.WHITE,
                ContextCompat.getColor(context, R.color.colorPressGrey))
    }

    /**
     * 打印调试 view tree

     * @param rootView
     * *
     * @param level
     */
    protected fun printViewTree(rootView: View, level: Int) {
        logSpace(level, getViewMsg(rootView))
        if (rootView is ViewGroup) {
            val vg = rootView
            for (i in 0..vg.childCount - 1) {
                val v = vg.getChildAt(i)
                printViewTree(v, level + 1)
            }
        }
    }

    protected fun logViewTree(view: View) {
        printViewTree(view, 0)
    }

    protected fun printActivityViewTree(activity: Activity) {
        val contentView = activity.findViewById(android.R.id.content)
        printViewTree(contentView, 0)
    }

    protected fun printActivityWindowViewTree(activity: Activity) {
        log("-------start----------")
        printViewTree(activity.window.decorView, 0)
        log("-------end----------")
    }

    protected val transparentDrawable: Drawable
        get() = ColorDrawable(Color.TRANSPARENT)

    protected val whiteDrawable: Drawable
        get() = ColorDrawable(Color.WHITE)

    protected fun getViewMsg(view: View): String {
        val className = view.javaClass.name
        val id = view.id
        var text = ""
        if (view is TextView) {
            val textView = view
            text = textView.text.toString()
            text += "(" + textView.hint + ")"
        }
        return "$className,$id,$text,$view"
    }

    private fun logSpace(count: Int, msg: String) {
        val sb = StringBuilder()
        for (j in 0..count - 1) {
            sb.append("----")
        }
        sb.append(msg)
        log(sb.toString())
    }

    protected fun log(view: View) {
        log(getViewMsg(view))
    }

    protected fun log(msg: String) {
        LogUtil.log(TAG + ":" + msg)
    }

    protected fun log(e: Throwable) {
        LogUtil.log(e)
    }

    protected fun logSuperClass(clazz: Class<*>) {
        if (clazz == Any::class.java) {
            return
        }
        log(clazz.name)
        logSuperClass(clazz.superclass)
    }

    @JvmOverloads protected fun logStackTraces(methodCount: Int = 15, methodOffset: Int = 3) {
        val trace = Thread.currentThread().stackTrace
        var level = ""
        log("---------logStackTraces start----------")
        for (i in methodCount downTo 1) {
            val stackIndex = i + methodOffset
            if (stackIndex >= trace.size) {
                continue
            }
            val builder = StringBuilder()
            builder.append("|")
                    .append(' ')
                    .append(level)
                    .append(trace[stackIndex].className)
                    .append(".")
                    .append(trace[stackIndex].methodName)
                    .append(" ")
                    .append(" (")
                    .append(trace[stackIndex].fileName)
                    .append(":")
                    .append(trace[stackIndex].lineNumber)
                    .append(")")
            level += "   "
            log(builder.toString())
        }
        log("---------logStackTraces end----------")
    }
}
